src/rust_connection/mod.rs

```diff
@@ -311,19 +311,19 @@ impl<R: ReadFD, W: WriteFD> RustConnection<R, W> {
                 let mut fds = Vec::new();
                 let mut packets = Vec::new();
 
-                let err = || -> std::io::Result<()> {
-                    // Read one packet
-                    mode.set_on_reader(&mut lock)?;
-                    packets.push(lock.read_packet(&mut fds)?);
+                // Read one packet
+                mode.set_on_reader(&mut lock)?;
+                if let Some(packet) = lock.try_read_packet(&mut fds)? {
+                    packets.push(packet);
 
                     // And then read as many packages as possible
                     BlockingMode::NonBlocking.set_on_reader(&mut lock)?;
-                    loop {
-                        packets.push(lock.read_packet(&mut fds)?);
+                    while let Some(packet) = lock.try_read_packet(&mut fds)? {
```

Comment:
This can lose the last successfully read packets in case of error. On single thread applications, these packets wouldn't be handled anyways because the function returns an error code. On multithread applications, it could be possible to allow the other thread handle the last packets, but then, once the thread tries to read again from the socket, it would also get an error.

Reply:
So you think that there are no temporary socket errors, great. That simplifies things.